### ISSUE TITLE
week2 earlier discussion of array/object cloning

### ISSUE THREAD

**Issue by rachaelcodes** [foundersandcoders/master-reference#740]
It may be worth adding a discussion of array/object cloning to the [pure functions workshop](https://github.com/foundersandcoders/pure-functions-easy-testing) (before students look at writing pure functions using these data types). 

The **Why do we need cloneArrayOfObjects?** discussion from the [project skeleton readme](https://github.com/foundersandcoders/master-reference/blob/master/coursebook/week-2/project/README.md) could be moved here instead.

**Comment by shannonjensen** [comment]
As a student who just went through this week, I agree. This [article](http://reactkungfu.com/2015/08/pros-and-cons-of-using-immutability-with-react-js/) has a very good simple way of explaining reference vs value equality - and a good analogy using identical cars.

Also, is there a reason we use cloneArrayOfObjects instead of just having students use array.slice() or [].concat(array)? Having cloneArrayOfObjects as its own function (and pre-written) excuses people from having to think about something quite crucial on their own. It would be better to have a reminder in the skeleton code that students must output a new array but not handholding so much (esp when that handholding seems to be done in an unnecessarily complicated way?)

**Comment by eliascodes** [comment]
> is there a reason we use cloneArrayOfObjects instead of just having students use array.slice() or [].concat(array)

That will just clone the array, not the objects in the array.

<hr>

tbh I agree with @shannonjensen that students should have to engage with the problem of object identity themselves. The todo project doesn't seem so large that it would be overwhelming with an extra problem thrown in. However, I think it wouldn't hurt to highlight the issue in the project documentation.

**Comment by ZooeyMiller** [comment]
I agree this is a problem perhaps students should learn to deal with on their own. In G3 not everyone used `cloneArrayOfObjects`, and I think that was because they didn't understand why they needed it.

Those that did use it spent a long time trying to understand it, and I think a more valuable use of that time would be for the students to work out their own implementation.

I think the problem may lay in how we make it clear what the difference between passing by reference and actually cloning an object is without just spoon feeding the answer. Especially when talking about arrays of objects. 

Perhaps we could add something onto the pure functions workshop as @shannonjensen mentioned? This was we would be able to write the tests, to ensure the students are forced to deal with the concept properly. 

I ended up giving a 20 min whiteboard walkthrough after code review for G3 to explain this function as it seemed to have been something not understood by a lot of the cohort.
